### DISCUSSION TITLE
Hot Reload is a Hot Mess - Bring Back Designer Please

### DISCUSSION THREAD

**Discussion by pouchbunny** [dotnet/maui#8652] [General]
### Description

Not sure if this is a Visual Studio issue or something directly related to MAUI extension.

The designer saved a lot of time when trying to place new controls and immediately told you if you broke something. Hot reload doesn't come close to what designer was. 

Working with hot reload in android is somewhat doable on a physical device, but there are plenty of issues with controls and elements not showing up or when changing and pressing hot reload does nothing.

However, in windows I get the following when trying to make any changes to a control:

### Steps to Reproduce

1. Create app
2. Run as windows app
3. Change one control (ie: font size or text)
4. First Exception thrown:  

System.IO.FileNotFoundException: 'Could not load file or assembly 'C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\PREVIEW\COMMON7\IDE\EXTENSIONS\XAMARIN\HOTRELOAD\Xamarin.HotReload.Translations.resources.dll'. The system cannot find the file specified.'

5. Second Exception thrown:

System.IO.FileNotFoundException: 'Could not load file or assembly 'C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\PREVIEW\COMMON7\IDE\EXTENSIONS\XAMARIN\HOTRELOAD\Xamarin.HotReload.Translations.resources.dll'. The system cannot find the file specified.'

6. Opens up and becomes your primary tabbed window:

C:\Users\<user>\AppData\Local\SourceServer\eadebd0e280cb63072d4f3dab4bae61499c95ec06ecba9ac0adcfe780209ab51\src\coreclr\System.Private.CoreLib\src\System\Resources\ManifestBasedResourceGroveler.CoreCLR.cs

7. Once you clear these steps you can then go back to making changes w/o new exceptions thrown, however when you reload the project and make changes the process starts all over.

Microsoft Visual Studio Professional 2022
Version 17.3.0 Preview 2.0
VisualStudio.17.Preview/17.3.0-pre.2.0+32611.2
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

ASP.NET and Web Tools   17.3.252.56319
ASP.NET and Web Tools

Azure App Service Tools v3.0.0   17.3.252.56319
Azure App Service Tools v3.0.0

C# Tools   4.3.0-2.22307.7+069a85a786acae3be2c7b49e330b7f968a624c88
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Mono Debugging for Visual Studio   17.3.11 (4f73250)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   6.3.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

Razor (ASP.NET Core)   17.0.0.2226702+6ac021dcee9db30337fdfc99f710c685002e5dc6
Provides languages services for ASP.NET Core Razor.

TypeScript Tools   17.0.10601.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   4.3.0-2.22307.7+069a85a786acae3be2c7b49e330b7f968a624c88
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   17.1.0-beta.22281.4+2b8fcc0c0a1cf1d69380159c7d03be993bba4016
Microsoft Visual F# Tools

Visual Studio IntelliCode   2.2
AI-assisted development for Visual Studio.

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

VSPackage Extension   1.0
VSPackage Visual Studio Extension Detailed Info

Xamarin   17.3.0.220 (main@1d7b676)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   17.3.0.102 (remotes/origin/d17-3@9d1ed6d99)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   17.2.15 (2e3b60e)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   12.3.99.95 (main/5432886)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: dffa5ab
    Java.Interop: xamarin/java.interop/main@1f27ab55
    SQLite: xamarin/sqlite/3.38.5@df4deab
    Xamarin.Android Tools: xamarin/xamarin-android-tools/main@fc3c2ac

Xamarin.iOS and Xamarin.Mac SDK   15.11.0.308 (7ad9f8893)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

### Version with bug

6.0.400 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Windows

### Affected platform versions

Windows 10

**Comment by janseris** [comment]
Maui needs a designer

**Comment by jsuarezruiz** [comment]
Can you share the Visual Studio version used?

**Reply by anpin** [reply]
seeing the same issue debugging windows target in this version: 
```
Microsoft Visual Studio Community 2022 (3)
Version 17.3.0 Preview 2.0
VisualStudio.17.Preview/17.3.0-pre.2.0+32611.2
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Community

Visual C++ 2022   00482-10000-00261-AA133
Microsoft Visual C++ 2022

ASP.NET and Web Tools   17.3.252.56319
ASP.NET and Web Tools

Azure App Service Tools v3.0.0   17.3.252.56319
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   17.3.252.56319
Azure Functions and Web Jobs Tools

C# Tools   4.3.0-2.22307.7+069a85a786acae3be2c7b49e330b7f968a624c88
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Entity Framework Core Power Tools   2.5
Adds useful design-time EF Core DbContext features to the Visual Studio Solution Explorer context menu.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Microsoft Azure Tools for Visual Studio   2.9
Support for Azure Cloud Services projects

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Mono Debugging for Visual Studio   17.3.11 (4f73250)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   6.3.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

Project System Tools   1.0
Tools for working with C#, VisualBasic, and F# projects.

Razor (ASP.NET Core)   17.0.0.2226702+6ac021dcee9db30337fdfc99f710c685002e5dc6
Provides languages services for ASP.NET Core Razor.

SQL Server Data Tools   17.0.62206.01190
Microsoft SQL Server Data Tools

Syntax Visualizer   1.0
An extension for visualizing Roslyn SyntaxTrees.

TypeScript Tools   17.0.10601.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   4.3.0-2.22307.7+069a85a786acae3be2c7b49e330b7f968a624c88
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   17.1.0-beta.22281.4+2b8fcc0c0a1cf1d69380159c7d03be993bba4016
Microsoft Visual F# Tools

Visual Studio IntelliCode   2.2
AI-assisted development for Visual Studio.

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

VSPackage Extension   1.0
VSPackage Visual Studio Extension Detailed Info

Xamarin   17.3.0.220 (main@1d7b676)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   17.3.0.102 (remotes/origin/d17-3@9d1ed6d99)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   17.2.15 (2e3b60e)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   12.3.99.95 (main/5432886)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: dffa5ab
    Java.Interop: xamarin/java.interop/main@1f27ab55
    SQLite: xamarin/sqlite/3.38.5@df4deab
    Xamarin.Android Tools: xamarin/xamarin-android-tools/main@fc3c2ac


Xamarin.iOS and Xamarin.Mac SDK   15.11.0.308 (7ad9f8893)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.
```

**Comment by ghost** [comment]
Hi @pouchbunny. We have added the "s/needs-info" label to this issue, which indicates that we have an open question for you before we can take further action. This issue will be closed automatically in 7 days if we do not hear back from you by then - please feel free to re-open it if you come back to this issue after that time.

**Reply by Xyncgas** [reply]
look at them pretending like it's not happening, like they don't use their product in the same way a working class people does

**Comment by drasticactions** [comment]
(I work on the XAML Hot Reload team, these are my opinions)

First, thank you for trying Hot Reload. For the Translation DLL error, that's being thrown if you allow all exceptions to be thrown within your applications. It doesn't affect Hot Reload itself from working. I would file that as an issue with VS Feedback though, since most likely the WinUI Hot Reload stack isn't inserting or loading that DLL when it should, or it may not actually exist. 

I would highly encourage you, whenever you hit an issue trying to hot reload, to try to replicate whatever you were trying to change in C# Code to see if you can invoke the change in general, without XAML Hot Reload (Ex. If you try to change the background color of a page with XAML Hot Reload and it doesn't work, try changing it via C# code behind.) If it doesn't work there, it's an MAUI bug. 

Lastly, I'm okay with talking about a MAUI Designer, but if you think Hot Reload is a mess you should think what the designer runs on top of, and if it would have the same issues.

**Reply by sqlguy6212** [reply]
Hot reload does not work for me either.  To replicate the problem is no problem but since nothing happens not sure what I would send

**Reply by danielftz** [reply]
> (I work on the XAML Hot Reload team, these are my opinions)
> 
> First, thank you for trying Hot Reload. For the Translation DLL error, that's being thrown if you allow all exceptions to be thrown within your applications. It doesn't affect Hot Reload itself from working. I would file that as an issue with VS Feedback though, since most likely the WinUI Hot Reload stack isn't inserting or loading that DLL when it should, or it may not actually exist.
> 
> I would highly encourage you, whenever you hit an issue trying to hot reload, to try to replicate whatever you were trying to change in C# Code to see if you can invoke the change in general, without XAML Hot Reload (Ex. If you try to change the background color of a page with XAML Hot Reload and it doesn't work, try changing it via C# code behind.) If it doesn't work there, it's an MAUI bug.
> 
> Lastly, I'm okay with talking about a MAUI Designer, but if you think Hot Reload is a mess you should think what the designer runs on top of, and if it would have the same issues.

Hot reload currently does not work with C# at all when Im using community toolkit markup. Issue here (https://developercommunity2.visualstudio.com/t/Hot-Reload-Not-Working-with-NET-MAUI-Co/10081168?space=8&q=dot+net+maui+hot+reload+markup&sort=relevance)

**Reply by gilyadid** [reply]
> Lastly, I'm okay with talking about a MAUI Designer, but if you think Hot Reload is a mess you should think what the designer runs on top of, and if it would have the same issues.

I understand your line of thought but I think you are missing the point here, even if the designer runs on the same execution principles being used for Hot-Reload, there is a clear difference in UX with those two-features. There are also many features that are specific to a designer like align, grids, snap to grid and others that are a bliss when working with a designer. 
By removing the designer support what is actually being removed is a quality grade for visual studio as IDE - and that is true for every designer being removed from the IDE, and yup, the EDMX as well.. IDE purpose is to give us UX which make our routine work easier and fun.

**Comment by janseris** [comment]
> (I work on the XAML Hot Reload team, these are my opinions)
> 
> First, thank you for trying Hot Reload. For the Translation DLL error, that's being thrown if you allow all exceptions to be thrown within your applications. It doesn't affect Hot Reload itself from working. I would file that as an issue with VS Feedback though, since most likely the WinUI Hot Reload stack isn't inserting or loading that DLL when it should, or it may not actually exist.
> 
> I would highly encourage you, whenever you hit an issue trying to hot reload, to try to replicate whatever you were trying to change in C# Code to see if you can invoke the change in general, without XAML Hot Reload (Ex. If you try to change the background color of a page with XAML Hot Reload and it doesn't work, try changing it via C# code behind.) If it doesn't work there, it's an MAUI bug.
> 
> Lastly, I'm okay with talking about a MAUI Designer, but if you think Hot Reload is a mess you should think what the designer runs on top of, and if it would have the same issues.

WPF designer exists in Visual Studio since at least 2008, that's 12 or 13 years before hot reload. 
Does MAUI designer have to be built on hot reload?

The designer feature is to design components individually without the context of the application

**Reply by sqlguy6212** [reply]
We need a designer Hot Reload just does nothing for me

**Reply by Xyncgas** [reply]
Maybe use what's in a human's skull to hot reload, maybe write codes that doesn't require debugging, the concept of business rules engine should be your enterprise internal framework that makes writing code like scripting that can't go wrong

**Comment by MagicAndre1981** [comment]
this was suggested some time ago #3439 and I agree, that we need a designer. This (XAML/C#) Hotreload also never worked for me

**Comment by xtuzy** [comment]
if xaml hotrealod not work, you can see my library ReloadPreview, it use reflection to load dll to preview, i used it when hotreload break.

**Comment by markadrake** [comment]
I frequently run my MAUI projects directly from my mac to the ios simulator. I suppose "hot reload" is only available if you use Visual Studio. It doesn't seem to be a feature I can use. 

Would anyone be kind enough to confirm hot reload is VS only? Thanks.

**Comment by pouchbunny** [comment]
Brand new app showing hot reload does not work.
https://user-images.githubusercontent.com/42586025/178281593-27d4b1ce-7d7b-4248-a5ac-e763da573ffa.mp4



> (I work on the XAML Hot Reload team, these are my opinions)
> 
> First, thank you for trying Hot Reload. For the Translation DLL error, that's being thrown if you allow all exceptions to be thrown within your applications. It doesn't affect Hot Reload itself from working. I would file that as an issue with VS Feedback though, since most likely the WinUI Hot Reload stack isn't inserting or loading that DLL when it should, or it may not actually exist.
> 
> I would highly encourage you, whenever you hit an issue trying to hot reload, to try to replicate whatever you were trying to change in C# Code to see if you can invoke the change in general, without XAML Hot Reload (Ex. If you try to change the background color of a page with XAML Hot Reload and it doesn't work, try changing it via C# code behind.) If it doesn't work there, it's an MAUI bug.
> 
> Lastly, I'm okay with talking about a MAUI Designer, but if you think Hot Reload is a mess you should think what the designer runs on top of, and if it would have the same issues.

Brand new app showing hot reload does not work by changing the background color in the code-behind.
https://user-images.githubusercontent.com/42586025/178281593-27d4b1ce-7d7b-4248-a5ac-e763da573ffa.mp4

**Comment by drasticactions** [comment]
@pouchbunny That's C# Hot Reload, and you're changing a value in the constructer. The only way that would change is if you reinvoked creating the page again.

**Reply by pouchbunny** [reply]
So Hot Reload is suppose to update the running app when it doesn't update once the change is made in the XAML? Does that include when using OnIdiom?

**Reply by Xyncgas** [reply]
Hot reload is not advanced enough to remember what you did to get to the point you are so if you changed something in a sub-routine that doesn't not propagate well into the side effects then into the memory of your objects but if you changed the value of say a variable that's referenced by other objects then the app is going to be showing that change.

They are having a hard moment developing hot reload that can trace all side effects, that's a general computer science problem and I guess no one is smart enough to figure it out yet, but if MAUI doesn't use XAML and Haskell was used instead of C# it may be possible to undo and redo steps that are executed since the application started therefore being able to trace how the changes in codes can affect the current state of the application

**Comment by drasticactions** [comment]
I'm moving this to discussions. There is nothing actionable on this issue that could be "fixed" here, but the discussion is a good one.

**Reply by davepruitt** [reply]
I absolutely agree with many of the people on this thread that we _need_ a designer. Hot Reload has _never_ been a satisfactory substitute for a XAML previewer/designer. WPF has had a previewer/designer for over a decade and it works marvelously. Xamarin Forms had a previewer/designer, and for some odd unknown reason Microsoft decided to get rid of it. Why?!?! It makes absolutely no sense.

Having a previewer/designer allows you do design and visualize your UI in _design time_. I _hate_ the fact that when using Hot Reload I have to run my program and navigate to the correct screen before I can get some kind of "preview" of my UI.

I would love to talk to whatever manager or team member at Microsoft made the decision to go "all in" on Hot Reload and not provide a previewer/designer, because it is a horrible mistake.

**Reply by davepruitt** [reply]
Also, it's clear that many people want this just by looking at the following other issues:

https://github.com/dotnet/maui/issues/3439
https://github.com/dotnet/maui/issues/6782

I'm also surprised that y'all keep closing these issues. Rather than close the issues - which gives us users the idea that you aren't taking any action on the issue - how about incorporating them into your to-do list?

**Reply by davepruitt** [reply]
My previous issue that I opened about this exact topic (Issue #6782 which I linked above) received 13 heart reactions, ranking it 14th in a list of nearly 5000 issues (both opened and closed) according to Github, and yet y'all just simply closed it. SMH.

**Reply by janseris** [reply]
> I'm moving this to discussions. There is nothing actionable on this issue that could be "fixed" here, but the discussion is a good one.

Instead of closing issues and ignoring the developers you should focus on solving it because people will move out of MAUI and it will simply die

**Reply by pmnlla** [reply]
I don't think this is a _discussion_ per se, I think it's an issue that's going to lead to a reduced market share for WinUI and Xamarin and would lead to people either returning to WPF or quitting native windows development all together.

I'm not a frontend developer. I'll never need be a frontend developer because I don't think I need to do frontend and I don't like to do frontend by itself. Or at least, I didn't need to go out of my way to learn to like frontend because it was practically effortless with frameworks like WPF and its designer view.

This is the kind of cheap shit that microsoft keeps pulling that makes me think to completely abandon the windows ecosystem and make something more progressive. Obviously not with WASM or that bloated chromium instance that discord, notion, tetrio, cider, and too many apps use that I refuse to show in a positive light because it's somehow managed to be more of a resource hog than UWP.
winUI isn't that progressive framework I'd love to use either. Nor is maui, or xamarin. It's an oblong mess of bodged-together DLLs from 1995, a coat of paint, and poor waste management. It's not the next generation of WPF: it's a shitshow.

PS: Seriously, why does tetris need an entire chrome instance to itself?

**Reply by Xyncgas** [reply]
Is this pompompuring speaking from prison

**Comment by tbrentlong** [comment]
Months and months later, and hot reload still does not work?!  I started a new project, and something as simple as changing the text on a label won't even refresh on the emulator. And why throw away the XAML designer?  We don't want to edit XAML as text.  Having to stop and restart to see every single change is ridiculous and incredibly time consuming.

**Comment by anpin** [comment]
Was it ever working? Never really worked for me since the time of xamarin. Really frustrating issue indeed.   On Thu, May 11, 2023 at 10:15 AM, Xyncgas ***@***.***> wrote:  
Maybe use what's in a human's skull to hot reload

**Reply by janseris** [reply]
In MAUI nothing than Xaml hot reload works for me in practice. Compared to a Windows Forms project where almost any code change can be hot reloaded. In MAUI, the percentage is close to zero cases.

**Reply by Donnyp751** [reply]
Same here, It's pretty much useless.

**Comment by WangYini** [comment]
hot reload doesnt work for me either when I'm using windows laptop with an ios local device. It is not possible with c# hot reload at all with this template MAUI project. And with xaml hot reload sometimes it works, sometimes it doesn't work. You don't know when this thing will throw several exceptions and then close the program.

**Comment by pmnlla** [comment]
Just gonna say something wild here to see if MS is actually overlooking this thread or just ignoring it all together, for shits and giggles. 

░M░Y░D░E░S░I░G░N░E░R░ ░V░I░E░W░ ░I░ N░B░I░O ░

**Comment by gautambjain** [comment]
Hot Reload is working pretty well for me. Saves a lot of time. 

If you are using it to design your view and controls, then it wouldn't work well for you. But you should use Hot Reload to "correct/improve" your view.

**Comment by developer9969** [comment]
Hot reload is a huge hit and miss - it works pretty well in hello world scenarios, as soon as you do something complex  real world than it has issues.. Designer is not the answer.. they just have to make work and stop using noddy sample in demos..

**Reply by gilyadid** [reply]
Designer is essential regardless of hot-reload, Hot-Reload is not a replacement for a designer.
Developers should be able to see the layout of views without executing the application.
The conceptual guideline is that any part of the application should be easy to locate and understand without executing it.
Designer is not just to "see", it also provide tools to edit the layout which you cannot have with just hot-reload.

**Comment by HannsKanns** [comment]
XAML Hot reload is still (June 2024) sometimes working and sometimes not as with many other MAUI issues. This could be a fundamental issue of MAUI open-source development with contributors. No contributor or employed Microsoft developer can earn his merits with the topics of reliability and professionalism in Buffixing topics.

**Comment by AlleSchonWeg** [comment]
Not only Hot Reload isn't reliable working. The toolchain is bad
- No C# Edit & Continue
- VS long path problem with iOS
- obj/bin folders needs sometimes manual delete
- VS for Mac abandoned. Maui plugin for VS Code is....
- Compiling with remote connection to a Mac is slow or results in errors
- iOS simulator for windows starts only sometimes. One day it works. The next day it stops working
- The app performance in Debug is slow. It makes it difficult to know if this your apps overall performance is good
- The app runs in debug fine, but in release it crashes. Then you have to start debug in release mode. 

To be fair, the problems also exits with Xamarin Forms. Personally i disabled Hot Reload, because it slows down the app and just reload XAML is mostly not enough. I need to reload data from the viewmodel and the layout after a Hot Reload looks different as you open the view normally.

**Reply by HannsKanns** [reply]
- Compiling a big code base with Android (Java) enabled is a nightmare and time-consuming. I only work with Windows and iOS; after some important milestones, I include Android for some necessary alignments.
- from my point of view, MAUI is still not production ready.
